server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -1902,11 +1906,10 @@ public boolean assertAllListenersResolved() {
                     } else {
                         final String updatedRepository = updateSnapshotState.snapshot().getRepository();
                         final Set<ShardId> reusedShardIds = reusedShardIdsByRepo.computeIfAbsent(updatedRepository, k -> new HashSet<>());
-                        if (entry.repository().equals(updatedRepository) &&
-                            entry.state().completed() == false && reusedShardIds.contains(finishedShardId) == false
-                                && entry.shards().keys().contains(finishedShardId)) {
+                        if (entry.state().completed() == false && entry.repository().equals(updatedRepository)
+                                && reusedShardIds.contains(finishedShardId) == false) {
                             final ShardSnapshotStatus existingStatus = entry.shards().get(finishedShardId);
-                            if (existingStatus.state() != ShardState.QUEUED) {
+                            if (existingStatus == null || existingStatus.state() != ShardState.QUEUED) {
```

Comment:
This is very hot code (shard state updates path) so it's worth saving one map lookup here IMO.